### PR TITLE
JAMES-3680 Allow Jmap session endpoint return dynamic prefix in response

### PR DESCRIPTION
**Pull request by vttranlina**
Jira: https://issues.apache.org/jira/browse/JAMES-3680 - OIDC setup for JMAP, IMAP, SMTP

As it stands now, the James session endpoint returns its host prefix.
```
}
	...
    "username": "bob@james.org",
    "apiUrl": "http://localhost/jmap",
    "downloadUrl": "http://localhost/download/{accountId}/{blobId}?type={type}&name={name}",
    "uploadUrl": "http://localhost/upload/{accountId}",
    "eventSourceUrl": "http://localhost/eventSource?types={types}&closeAfter={closeafter}&ping={ping}",
    "state": "2c9f1b12-b35a-43e6-9af2-0106fb53a943"
}
```
With 2 Krakend endpoints (normal auth one and OIDC one), Jmap session still return the same its url prefix. User would be bias by this response and think he can use same authentication strategy for 2 endpoints.
Beside that with API gateway standing in front of a backend, we should hide the behind backend information.

## How
- Define a configuration pair in `jmap.properties` file allow dynamic prefix response
	```
	dynamic.jmap.prefix.resolution.enabled=true
	```
- Add a field in `JmapRfc8621Configuration` to store this
- When this config is true and X-JMAP-Prefix header is passed from Krakend to JAMES' session endpoint, we should modify prefix in response following X-JMAP-Prefix header's value. We can do this by adding a header filter when handling response for session route.

### CODE REVIEW

**Comment by vttranlina** [comment]
Should we set dynamic url prefix for `webSocketUrl`?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala [hunk 1]
```diff
@@ -23,45 +23,65 @@ import java.net.{URI, URL}
...
+    .getOrElse(urlPrefixString)
+
+  def apiUrl(urlPrefixRequest: Option[URL] ): URL = new URL(s"${getURLPrefix(urlPrefixRequest)}/jmap")
+
+  def downloadUrl(urlPrefixRequest: Option[URL]): URL = new URL(s"${getURLPrefix(urlPrefixRequest)}/download/{accountId}/{blobId}?type={type}&name={name}")
+
+  def uploadUrl(urlPrefixRequest: Option[URL]): URL = new URL(s"${getURLPrefix(urlPrefixRequest)}/upload/{accountId}")
+
+  val webSocketUrl: URI = new URI(s"$websocketPrefixString/jmap/ws")
+  val eventSourceUrl: URL = new URL(s"$urlPrefixString/eventSource?types={types}&closeAfter={closeafter}&ping={ping}")
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala:84]: Is a configuration POJO the right place for advanced resolution? Shouldn't an other entity do this logic relying only on the prefix shiped by the configuration?

**Review by chibenwa — COMMENTED** [review]
IMO we need a clear statement within this PR stating why this PR is useful.

ALso we need a test for the session when `X-JMAP-Prefix` header is used, and do not forget documentation.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala [hunk 2]
```diff
@@ -43,33 +44,44 @@ object JmapRfc8621Configuration {
...
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala:84]: Don't forget to resolve `webSocketUrl` and `eventSourceUrl` as well

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Mistake I do not approve sorry :)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala (hunk 2 above)
Comment (vttranlina) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala:84]: Then we need one more header to set WebSocket url prefix. I'm right?

1. X-JMAP-PREFIX:
- https://domain.tld/oidc
- http://domain.tld/oidc

2. X-JMAP-WEBSOCKET-PREFIX
- wss://domain.tld/oidc
- ws://domain.tld/oidc

?

Comment (vttranlina) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala:84]: I figure out, 
X-JMAP-PREFIX should have only domain and path, NOT contain schema. 
Then we can use it to replace in both urlPrefix and websocketUrlPrefix

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala (hunk 1 above)
Comment (vttranlina) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala]: I created the `JmapUrlEndpointResolver` class for this

File: server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc [hunk 3]
```diff
@@ -31,36 +31,37 @@ This should not be the same keystore than the ones used by TLS based protocols.
...
 | Optional string. Defaults to draft. Allowed values: draft, rfc-8621.
-| Which version of the JMAP protocol should be served when none supplied in the Accept header.
+Which version of the JMAP protocol should be served when none supplied in the Accept header.
 Defaults to draft for legacy reasons (avoid breaking changes) but setting the value to
 rfc-8621 allow compatibility with other third party apps.
 
+| dynamic.jmap.prefix.resolution.enabled
+| Optional boolean. Defaults to false. Supported Jmap session endpoint return dynamic prefix in response.
+When it was config is true, and http request to jmap session endpoint has `X-JMAP-PREFIX` header with value is `/jmapprefix`, then `apiUrl` in response will be "${url.prefix}/jmapprefix/jmap".
+Example: http://localhost/jmapprefix/jmap
```
Comment (chibenwa) [inline server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc:63]: IMO we should be more generic and override the url.prefix.

Eg:

`X-JMAP-PREFIX: http://sub.domainb.tld/prefix` then `apiUrl` in response will be "http://sub.domainb.tld/prefix/jmap".

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala [hunk 4]
```diff
@@ -171,4 +171,111 @@ trait SessionRoutesContract {
 
     assertThatJson(sessionJson).isEqualTo(expected_session_object)
   }
+
+  @Test
+  def getResponseShouldReturnUrlEndpointContainJmapPrefixWhenAssignHeaderAndConfigurationProvided(): Unit = {
+    val sessionJson: String = `given`()
+    .when()
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .header("X-JMAP-PREFIX", "/oidc")
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:180]: IMO we should have here a fully qualified URL. This gives greater flexibility when configuring an API gateway.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala [hunk 5]
```diff
@@ -171,4 +171,111 @@ trait SessionRoutesContract {
...
+        |    "urn:apache:james:params:jmap:mail:quota": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+        |    "urn:apache:james:params:jmap:mail:shares": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+        |    "urn:ietf:params:jmap:vacationresponse": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+        |    "urn:ietf:params:jmap:mdn": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6"
+        |  },
+        |  "username" : "bob@domain.tld",
+        |  "apiUrl" : "http://domain.com/oidc/jmap",
+        |  "downloadUrl" : "http://domain.com/oidc/download/{accountId}/{blobId}?type={type}&name={name}",
+        |  "uploadUrl" : "http://domain.com/oidc/upload/{accountId}",
+        |  "eventSourceUrl" : "http://domain.com/oidc/eventSource?types={types}&closeAfter={closeafter}&ping={ping}",
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:276]: Maybe we could find a way to only assert those URLs ?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala (hunk 2 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala]: the X-JMAP-PREFIX is the same for every endpoint, it's unrelated to urlPrefix and websocketUrlPrefix

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala (hunk 4 above)
Comment (vttranlina) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:180]: What do you think about this?
https://github.com/apache/james-project/pull/780#discussion_r763726881

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala (hunk 2 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala]: I think the two header approach is valid.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala (hunk 4 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:180]: I think the two header approach is valid.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala [hunk 6]
```diff
@@ -174,108 +175,37 @@ trait SessionRoutesContract {
 
   @Test
   def getResponseShouldReturnUrlEndpointContainJmapPrefixWhenAssignHeaderAndConfigurationProvided(): Unit = {
-    val sessionJson: String = `given`()
+    `given`()
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .header("X-JMAP-PREFIX", "http://new-domain.tld/prefix")
+      .header("X-JMAP-WEBSOCKET-PREFIX", "ws://new-domain.tld/prefix")
     .when()
+      .get("/session").prettyPeek()
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:183]: debug spotted

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/SessionSupplier.scala [hunk 7]
```diff
@@ -68,20 +68,27 @@ class SessionSupplier(val configuration: JmapRfc8621Configuration, defaultCapabi
       .toMap
 }
 
-class JmapUrlEndpointResolver(val configuration: JmapRfc8621Configuration, jmapPrefixRequest: Option[String] = None) {
+class JmapUrlEndpointResolver(val configuration: JmapRfc8621Configuration,
+                              urlPrefixRequest: Option[URL] = None,
+                              uriWebSocketPrefixRequest: Option[URI]= None) {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/SessionSupplier.scala:73]: space between = amd ]

**Review by chibenwa — APPROVED** [review]

File: server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc [hunk 8]
```diff
@@ -31,36 +31,39 @@ This should not be the same keystore than the ones used by TLS based protocols.
...
 
 | jmap.version.default
 | Optional string. Defaults to draft. Allowed values: draft, rfc-8621.
-| Which version of the JMAP protocol should be served when none supplied in the Accept header.
+Which version of the JMAP protocol should be served when none supplied in the Accept header.
 Defaults to draft for legacy reasons (avoid breaking changes) but setting the value to
 rfc-8621 allow compatibility with other third party apps.
 
+| dynamic.jmap.prefix.resolution.enabled
+| Optional boolean. Defaults to false. Supported Jmap session endpoint return dynamic prefix in response.
```
Comment (Arsnael) [inline server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc:61]:
```suggestion
| Optional boolean. Defaults to false. Supported Jmap session endpoint returns dynamic prefix in response.
```

File: server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc [hunk 9]
```diff
@@ -31,36 +31,39 @@ This should not be the same keystore than the ones used by TLS based protocols.
...
+When it was config is true, and http request to jmap session endpoint has `X-JMAP-PREFIX` header with value is `http://new-domain/prefix`,
```
Comment (Arsnael) [inline server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc:62]:
```suggestion
When its config is true, and the HTTP request to Jmap session endpoint has a `X-JMAP-PREFIX` header with the value `http://new-domain/prefix`,
```

File: server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc [hunk 10]
```diff
@@ -31,36 +31,39 @@ This should not be the same keystore than the ones used by TLS based protocols.
...
+then `apiUrl` in response will be "http://new-domain/prefix/jmap".
```
Comment (Arsnael) [inline server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc:63]: only `apiUrl` is concerned?

File: server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc [hunk 11]
```diff
@@ -31,36 +31,39 @@ This should not be the same keystore than the ones used by TLS based protocols.
...
+If http request to jmap session endpoint has `X-JMAP-WEBSOCKET-PREFIX` header with value is `ws://new-domain/prefix`,
```
Comment (Arsnael) [inline server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmap.adoc:64]:
```suggestion
If the HTTP request to Jmap session endpoint has the `X-JMAP-WEBSOCKET-PREFIX` header with the value `ws://new-domain/prefix`,
```

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/SessionRoutes.scala [hunk 12]
```diff
@@ -15,10 +15,11 @@
  * KIND, either express or implied.  See the License for the    *
  * specific language governing permissions and limitations      *
  * under the License.                                           *
- * ***************************************************************/
+ * ************************************************************** */
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/SessionRoutes.scala:18]: Format error

File: src/site/xdoc/server/config-jmap.xml [hunk 13]
```diff
@@ -89,6 +89,14 @@
                     <dd>Which version of the JMAP protocol should be served when none supplied in the Accept header.
                     Defaults to draft for legacy reasons (avoid breaking changes) but setting the value to
                     rfc-8621 allow compatibility with other third party apps.</dd>
+
+                    <dt><strong>dynamic.jmap.prefix.resolution.enabled</strong></dt>
+                    <dd>Optional boolean. Defaults to false. Supported Jmap session endpoint return dynamic prefix in response.
```
Comment (Arsnael) [inline src/site/xdoc/server/config-jmap.xml:94]: Rewrite that to my suggestions on the same text in the doc commented above please :)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala [hunk 14]
```diff
@@ -19,49 +19,57 @@
...
                                     maxTimeoutSeconds: Option[Int] = None,
                                     maxConnections: Option[Int] = None) {
-  val urlPrefix: URL = new URL(urlPrefixString)
-  val apiUrl: URL = new URL(s"$urlPrefixString/jmap")
-  val downloadUrl: URL = new URL(urlPrefixString + "/download/{accountId}/{blobId}?type={type}&name={name}")
-  val uploadUrl: URL = new URL(s"$urlPrefixString/upload/{accountId}")
-  val eventSourceUrl: URL = new URL(s"$urlPrefixString/eventSource?types={types}&closeAfter={closeafter}&ping={ping}")
-  val webSocketUrl: URI = new URI(s"$websocketPrefixString/jmap/ws")
 
+  val webSocketUrl: URI = new URI(s"$websocketPrefixString/jmap/ws")
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala:72]: As you resolve this in your `JmapUrlEndpointResolver` now, this line is not necessary anymore?

Comment (vttranlina) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala:72]: This line still is invoked some where.
 (`org.apache.james.jmap.core.DefaultCapabilities#supported` , `org.apache.james.jmap.draft.JMAPModule#webSocketCapability`)
 - `JmapUrlEndpointResolver` just change the HTTP response case.

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala:72]: Shouldn't those be adapted to the new url endpoint resolution as well then?

**Review by Arsnael — APPROVED** [review]

**Comment by vttranlina** [comment]
rebase master & squash fixup

**Pull request merged**